File: problems/590-n-ary-tree-postorder-traversal/stack.js
/**
 * 循环法：最简单的stack思路
 * 
 * 时间：716ms
 */
var postorder = function(root) {
  const res = []
  if (!root) return res

  const stack = new Stack()
  stack.push(root)

  while (!stack.empty()) {
    const curr = stack.pop()

    // 遍历本结点
    res.push(curr.val)

    // 顺序压栈
    for (let i = 0; i < curr.children.length; ++i) {
      stack.push(curr.children[i])
    }
  }

  return res.reverse()
};

class Stack {
  constructor (arr = []) {
    this.arr = arr
  }

  push (n) {
    this.arr.push(n)
  }

  pop () {
    return this.arr.pop()
  }

  empty () {
    return this.arr.length === 0
  }

  top () {
    return this.arr[this.arr.length - 1]
  }
}

File: problems/439-ternary-expression-parser/-stack.cpp
class Solution {
public:
    string parseTernary(string S) {
        const int n = S.size();
        char curr = EMPTY;

        for (int i = 0; i < n; ++i) {
            if (curr == EMPTY) {
                curr = S[i];
                continue;
            }

            if (S[i] == '?') {
                conds.push(curr);
            } else if (S[i] == ':') {
                ops.push(curr);
            }

            if (!conds.empty() && ops.size() >= 2) {
                curr = getAndEvaluate();
            } else {
                curr = EMPTY;
            }
        }
        ops.push(curr);

        while (!conds.empty()) {
            ops.push(getAndEvaluate());
        }

        string res;
        res.push_back(ops.top());
        return res;
    }

    char getAndEvaluate() {
        char cond = conds.top();
        conds.pop();
        char b = ops.top();
        ops.pop();
        char a = ops.top();
        ops.pop();
        return cond == 'T' ? a : b;
    }

private:
    const char EMPTY = ' ';
    stack<char> conds;
    stack<char> ops;
};